### PR TITLE
0009178: Fix bindKey & unbindKey on commands

### PR DESCRIPTION
**Pull request by emre1702**
Bugtracker: 
https://bugs.mtasa.com/view.php?id=9178

bindKey and unbindKey were really strange for commands, so I rewrote them a bit. 
Now it should work perfectly (tested it as much as I could, no problems for me).

Still wonder why "SetAllCommandsActive" was used there.

You can ignore the first commit (f1c3e50), wasn't the solution.

### CODE REVIEW

**Comment by emre1702** [comment]
You can ignore the first commit (f1c3e50), wasn't the solution.

**Comment by ccw808** [comment]
Thanks for the PR. Unfortunately, there are still problems:
```
bindKey("t", "down", "chatbox", "test1");
unbindKey("t", "down", "chatbox", "test1");

bindKey("t", "down", "chatbox", "test2");
unbindKey("t", "down", "chatbox", "test2");  -- test2 will not be unbound
```

**Comment by emre1702** [comment]
I already tested the code and it worked for me.
Did you test it and got the problem?

Edit:
Wth yes, now getting the same problem.
Sry, will work at it again and fix it.

**Comment by emre1702** [comment]
@ccw808 Found the problem:
SetCommandActive (de)activates the first bind it finds. 
With bindKey we add new binds, with unbindKey we only deactivate them, not delete.
So if we use unbindKey without an argument, SetCommandActive deactivates the first bind (again).

I could also have used SetAllCommandsActive, but it got no return. 
Using SetCommandActive and SetAllCommandsActive like it was before would iterate twice (in worst case), so isn't better.

Edit:
Please wait a bit before doing something, still fixing a small bug.

**Comment by emre1702** [comment]
Ok, now it should hopefully work.

File: Client/core/CKeyBinds.cpp [hunk 1]
```diff
@@ -771,6 +771,7 @@ bool CKeyBinds::SetCommandActive(const char* szKey, const char* szCommand, bool
     NullEmptyStrings(szKey, szCommand, szArguments);
 
     list<CKeyBind*>::const_iterator iter = m_pList->begin();
+    bool ret = false;
```
Comment (ccw808) [inline Client/core/CKeyBinds.cpp]: bReturn

File: Client/core/CKeyBinds.cpp [hunk 2]
```diff
@@ -784,18 +785,22 @@ bool CKeyBinds::SetCommandActive(const char* szKey, const char* szCommand, bool
                     {
                         if (!checkHitState || (pBind->bHitState == bState))
                         {
-                            if (!szArguments || (pBind->szArguments && strcmp(pBind->szArguments, szArguments) == 0))
+                            if (!szArguments && (!bActive || !pBind->szArguments) || szArguments && (pBind->szArguments && strcmp(pBind->szArguments, szArguments) == 0))
```
Comment (ccw808) [inline Client/core/CKeyBinds.cpp]: Needs more parentheses

File: Client/core/CKeyBinds.cpp [hunk 3]
```diff
@@ -784,18 +785,22 @@ bool CKeyBinds::SetCommandActive(const char* szKey, const char* szCommand, bool
                     {
                         if (!checkHitState || (pBind->bHitState == bState))
                         {
-                            if (!szArguments || (pBind->szArguments && strcmp(pBind->szArguments, szArguments) == 0))
+                            if (!szArguments && (!bActive || !pBind->szArguments) || szArguments && (pBind->szArguments && strcmp(pBind->szArguments, szArguments) == 0))
                             {
+                                bool oldval = pBind->bActive;
```
Comment (ccw808) [inline Client/core/CKeyBinds.cpp]: bOldActive

File: Client/core/CKeyBinds.cpp [hunk 4]
```diff
@@ -784,18 +785,22 @@ bool CKeyBinds::SetCommandActive(const char* szKey, const char* szCommand, bool
...
-                            if (!szArguments || (pBind->szArguments && strcmp(pBind->szArguments, szArguments) == 0))
+                            if (!szArguments && (!bActive || !pBind->szArguments) || szArguments && (pBind->szArguments && strcmp(pBind->szArguments, szArguments) == 0))
                             {
+                                bool oldval = pBind->bActive;
                                 pBind->bActive = bActive;
-                                return true;
+                                if (szArguments)
+                                    return true;
+                                else
+                                    ret = ret || oldval != bActive;
```
Comment (ccw808) [inline Client/core/CKeyBinds.cpp]: Needs more parentheses

**Comment by ccw808** [comment]
Works here. Just need some source tweaks

**Comment by emre1702** [comment]
@ccw808 Done

**Comment by ccw808** [comment]
Thanks - Good work.

**Pull request merged**